File: src/xewma_arl.c
#include <stdio.h>
#include <math.h>
#include <stdlib.h>
#include <R.h>

#define ewma1 0
#define ewma2 1
#define fix 0
#define vacl 1
#define fir 2
#define both 3
#define steiner 4
#define stat 5
#define fink 6
#define elimit 7
#define waldmann 8
#define collocation 9

extern double rho0;

double *vector (long n);
double xe1_iglarl(double l, double c, double zr, double hs, double mu, int N);
double xe1_arlm(double l, double c, double zr, double hs, int q, double mu0, double mu1,int mode, int N, int nmax);
double xe1_arlm_hom(double l, double c, double zr, double hs, int q, double mu0, double mu1, int N, double *ced);
double xlimit1_arlm(double c, double zr, int q, double mu0, double mu1, int N, int nmax);
double xe1_Warl(double l, double c, double zr, double hs, double mu, int N, int nmax);

double xe2_iglarl(double l, double c, double hs, double mu, int N);
double xe2_Warl(double l, double c, double hs, double mu, int N, int nmax);
double xe2_Carl(double l, double c, double hs, double mu, int N, int qm);
double xe2_arlm(double l, double c, double hs, int q, double mu0, double mu1, int mode, int N, int nmax);
double xe2_arlm_hom(double l, double c, double hs, int q, double mu0, double mu1, int N, double *ced);

void xewma_arl(int *ctyp, double *l, double *c, double *zr, double *hs, double *mu, int *ltyp, int *r, int *q, double *arl)
{ int nmax=100000, i, result=0;
  double *ced, arl1=-1.;
 ced  = vector(*q);

 if (*ctyp==ewma1 && *ltyp==fix && *q==1)
                     arl1 = xe1_iglarl(*l,*c,*zr,*hs,*mu,*r);
 if (*ctyp==ewma1 && *ltyp==fix && *q>1)
                     result = xe1_arlm_hom(*l, *c, *zr, *hs, *q, 0., *mu, *r, ced);
                     /* *arl = xe1_arlm(*l,*c,*zr,*hs,*q,0.,*mu,*ltyp,*r,nmax);*/
 if (*ctyp==ewma1 && *ltyp>fix && *ltyp<elimit)
                     arl1 = xe1_arlm(*l,*c,*zr,*hs,*q,0.,*mu,*ltyp,*r,nmax);
 if (*ctyp==ewma1 && *ltyp==elimit)
                     arl1 = xlimit1_arlm(*c,*zr,*q,0.,*mu,*r,nmax);
 if (*ctyp==ewma1 && *ltyp==waldmann)
                     arl1 = xe1_Warl(*l,*c,*zr,*hs,*mu,*r,nmax);

 if (*ctyp==ewma2 && *ltyp==fix && *q==1)
                     arl1 = xe2_iglarl(*l,*c,*hs,*mu,*r);
 if (*ctyp==ewma2 && *ltyp==fix && *q>1)
                     result = xe2_arlm_hom(*l, *c, *hs, *q, 0., *mu, *r, ced);                 
                     /* arl1 = xe2_arlm(*l,*c,*hs,*q,0.,*mu,*ltyp,*r,nmax);*/
 if (*ctyp==ewma2 && *ltyp>fix && *ltyp<waldmann)
                     arl1 = xe2_arlm(*l,*c,*hs,*q,0.,*mu,*ltyp,*r,nmax);
 if (*ctyp==ewma2 && *ltyp==waldmann)
                     arl1 = xe2_Warl(*l,*c,*hs,*mu,*r,nmax);
 if (*ctyp==ewma2 && *ltyp==collocation)
                     arl1 = xe2_Carl(*l,*c,*hs,*mu,*r,50);
 
 if ( result != 0 ) warning("trouble in xewma_arl [package spc]");
 
 if ( *ltyp==fix && *q>1 ) for (i=0; i<*q; i++) arl[i] = ced[i]; else *arl = arl1;
}

File: R/xewma.arl.R
# Computation of EWMA ARLs (mean monitoring)
xewma.arl <- function(l, cE, mu, zr=0, hs=0, sided="one", limits="fix", q=1, steady.state.mode="conditional", r=40) {
  if ( l<=0 | l>2 )
    stop("l has to be between 0 and 2")
  if ( any(cE<=0) )
    warning("usually, cE has to be positive")
  if ( limits!="cfar" ) {  
    if ( zr>cE & sided=="one" )
      stop("wrong reflexion border")
    if ( (sided=="two" & abs(hs)>cE) | (sided=="one" & (hs<zr | hs>cE)) ) 
      warning("unusual headstart")
  }
  if ( r<4 )
    stop("r is too small")
  ctyp <- pmatch(sided, c("one", "two")) - 1
  if ( is.na(ctyp) )
    stop("invalid ewma type")
  ltyp <- -1 + pmatch(limits,
          c("fix", "vacl", "fir", "both", "Steiner", "stat", "cfar", "limit", "fixW", "fixC"))            
  if ( is.na(ltyp) )
    stop("invalid limits type")
  if ( (sided=="one") & !(limits %in% c("fix", "vacl", "stat", "limit", "fixW")) )
    stop("not supported for one-sided EWMA (not reasonable or not implemented yet")
  q <- round(q)
  if ( q<1 )
    stop("wrong change point position (q)")
  styp <- pmatch(steady.state.mode, c("conditional", "cyclical")) - 1
  if (is.na(styp))
    stop("invalid steady.state.mode")
  if ( limits=="cfar" ) {
    ctyp <- length(cE) # nc
    hs <- cE[length(cE)] # cinf
  }
  if ( limits=="fix" & q>1 & styp==0 ) {
    arl <- .C("xewma_arl",as.integer(ctyp),as.double(l),
              as.double(cE),as.double(zr),as.double(hs),
              as.double(mu),as.integer(ltyp),as.integer(r),as.integer(q),as.integer(styp),
              ans=double(length=q), PACKAGE="spc")$ans 
  } else {
    arl <- .C("xewma_arl",as.integer(ctyp),as.double(l),
              as.double(cE),as.double(zr),as.double(hs),
              as.double(mu),as.integer(ltyp),as.integer(r),as.integer(q),as.integer(styp),
              ans=double(length=1), PACKAGE="spc")$ans
  }
  names(arl) <- NULL
  return (arl)
}
